recipes/PeoplesSpeech/ASR/transformer/train.py

```diff
@@ -0,0 +1,460 @@
+        "wrd", "tokens_list", "tokens_bos", "tokens_eos", "tokens"
+    )
+    def text_pipeline(wrd):
+
+        wrd = english_specific_preprocess(wrd)
+        yield wrd
+        tokens_list = tokenizer.sp.encode_as_ids(wrd)
+        yield tokens_list
+        tokens_bos = torch.LongTensor([hparams["bos_index"]] + (tokens_list))
+        yield tokens_bos
+        tokens_eos = torch.LongTensor(tokens_list + [hparams["eos_index"]])
+        yield tokens_eos
+        tokens = torch.LongTensor(tokens_list)
+        yield tokens
+
+    sb.dataio.dataset.add_dynamic_item(datasets, text_pipeline)
+
+    # 4. Set output:
+    sb.dataio.dataset.set_output_keys(
+        datasets,
+        ["id", "audio_id", "sig", "wrd", "tokens_bos", "tokens_eos", "tokens"],
+    )
+
+    # 5. If Dynamic Batching is used, we instantiate the needed samplers.
+    train_batch_sampler = None
+    valid_batch_sampler = None
+
+    dynamic_hparams = hparams["dynamic_batch_sampler"]
+    dynamic_hparams_valid = hparams["dynamic_batch_sampler_valid"]
+    num_buckets = dynamic_hparams["num_buckets"]
+
+    train_batch_sampler = DynamicBatchSampler(
+        train_data,
+        dynamic_hparams["max_batch_length"]
+        * 1000,  # duration is in ms so back to s.
+        num_buckets=num_buckets,
+        lengths_list=train_len_list,
+        shuffle=dynamic_hparams["shuffle"],
+        batch_ordering=dynamic_hparams["batch_ordering"],
+    )
+
+    valid_batch_sampler = DynamicBatchSampler(
+        valid_data,
+        dynamic_hparams_valid["max_batch_length"]
+        * 1000,  # duration is in ms so back to s.
+        num_buckets=num_buckets,
+        lengths_list=val_len_list,
+        shuffle=dynamic_hparams_valid["shuffle"],
+        batch_ordering=dynamic_hparams_valid["batch_ordering"],
+    )
+
+    return (
+        train_data,
+        valid_data,
+        test_data,
+        train_batch_sampler,
+        valid_batch_sampler,
+    )
+
+
+if __name__ == "__main__":
+    # CLI:
+    hparams_file, run_opts, overrides = sb.parse_arguments(sys.argv[1:])
+    with open(hparams_file, encoding="utf-8") as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    # create ddp_group with the right communication protocol
+    sb.utils.distributed.ddp_init_group(run_opts)
+
+    # Create experiment directory
+    sb.create_experiment_directory(
+        experiment_directory=hparams["output_folder"],
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
+    # Dataset download and preparation
+    from peoples_speech_prepare import (
+        english_specific_preprocess,
+        load_and_concatenate_datasets,
+        prepare_peoples_speech,
+    )
+
+    # multi-gpu (ddp) save data preparation
+    run_on_main(
+        prepare_peoples_speech,
+        kwargs={
+            "hf_download_folder": hparams["hf_download_folder"],
+            "subsets": hparams["subsets"],
+            "save_folder": hparams["save_folder"],
```

Comment:
Question: I found many recipes saving `csv` directly inside of the `output_folder`. In this case, you made the choice of saving manifests in the `save_folder` (i.e. `output_folder/save`). I was wondering, but, which one should we stick to in the future?

Reply:
Yes, it's a long standing issue, I don't have an answer for that to be honest..